User custom fields in the invite user modal

b12f opened the pull request:
This adds the dynamic form fields for required custom fields for new users. Happy path is done, tests are missing.

aleixsuau committed:
Projects form working with formly 50%

aleixsuau committed:
Projects form working with formly 50%

aleixsuau committed:
Removed console.log

aleixsuau committed:
Removed console.log

aleixsuau committed:
Working with formattable

aleixsuau committed:
Working with formattable

aleixsuau committed:
Working with formattable

aleixsuau committed:
Working with formattable

aleixsuau committed:
Input with id and label

aleixsuau committed:
Input with id and label

aleixsuau committed:
Input with id and label

aleixsuau committed:
Input with id and label

aleixsuau committed:
Useless dependencies removed

aleixsuau committed:
Useless dependencies removed

aleixsuau committed:
Saving forms + required labels with *

aleixsuau committed:
Saving forms + required labels with *

aleixsuau committed:
First backend validation approach

aleixsuau committed:
First backend validation approach

aleixsuau committed:
Removed reload on type change + keep model on route changes

aleixsuau committed:
Removed reload on type change + keep model on route changes

aleixsuau committed:
Handlig backend validations with setError

aleixsuau committed:
Handlig backend validations with setError

aleixsuau committed:
Formatting the form model to submit

aleixsuau committed:
Formatting the form model to submit

aleixsuau committed:
Make up refactor

aleixsuau committed:
Make up refactor

aleixsuau committed:
working with op-form-field

aleixsuau committed:
working with op-form-field

aleixsuau committed:
Form creation moved to the service

aleixsuau committed:
Form creation moved to the service

aleixsuau committed:
Working with op-form-field wrapper

aleixsuau committed:
Working with op-form-field wrapper

aleixsuau committed:
Working with validation and op-form-field

aleixsuau committed:
Working with validation and op-form-field

aleixsuau committed:
Working with []CustomFields

aleixsuau committed:
Working with []CustomFields

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Clean up

aleixsuau committed:
Form routing working

aleixsuau committed:
Form routing working

aleixsuau committed:
Notification on form error and success

aleixsuau committed:
Notification on form error and success

aleixsuau committed:
Refactor + removed useless dynamic form observable

aleixsuau committed:
Refactor + removed useless dynamic form observable

aleixsuau committed:
DynamicFieldsService with tests

aleixsuau committed:
DynamicFieldsService with tests

aleixsuau committed:
Refactor: inputs catalog + catch form load error

aleixsuau committed:
Refactor: inputs catalog + catch form load error

aleixsuau committed:
Filter out non writable fields

aleixsuau committed:
Filter out non writable fields

aleixsuau committed:
Refactor: naming consistency

aleixsuau committed:
Refactor: naming consistency

aleixsuau committed:
Cleaning comments

aleixsuau committed:
Cleaning comments

aleixsuau committed:
dynamic-fields-service tests + wrapper component

aleixsuau committed:
dynamic-fields-service tests + wrapper component

aleixsuau committed:
DynamicForm Tests

aleixsuau committed:
DynamicForm Tests

aleixsuau committed:
Removed projects routes and ruby template

aleixsuau committed:
Cleaning up

aleixsuau committed:
Cleaning up

aleixsuau committed:
Removed projects routes and dynamic forms from Projects

aleixsuau committed:
Provide DynamicFieldsService in root so it can be used independently

aleixsuau committed:
Revert "Provide DynamicFieldsService in root so it can be used independently"

This reverts commit ab56f3c56f6c8c3c3b791c081c762aa713e83a41.

aleixsuau committed:
Provide DynamicFieldsService in root so it can be used independently

aleixsuau committed:
TODO: test ProjectsComponent

aleixsuau committed:
Code climate fix

aleixsuau committed:
Code climate fix

aleixsuau committed:
Code climate fixes (remove TODOs)

aleixsuau committed:
Default OpFormFieldComponent.inlineLabel to false

aleixsuau committed:
@ngx-formly/core dependency added

aleixsuau committed:
@ngx-formly/core dependency added

aleixsuau committed:
Dynamic components tests xkipped

aleixsuau committed:
Removed CdkTextareaAutosize because of CDK issue 22469

aleixsuau committed:
Removed CdkTextareaAutosize because of CDK issue 22469

aleixsuau committed:
Typing improvements

aleixsuau committed:
DynamicForm working as a FormControl

aleixsuau committed:
DynamicForm working as a FormControl

aleixsuau committed:
DynamicFormComponent working as a FormControl

aleixsuau committed:
Getting route params sync

aleixsuau committed:
Getting route params sync

aleixsuau committed:
Global FormsService: submit + formatting + error handling

aleixsuau committed:
Global FormsService: submit + formatting + error handling

aleixsuau committed:
Global FormsService: submit + formatting + error handling

aleixsuau committed:
Fix: @Optional() FormGroupDirective in OpFormFieldComponent

aleixsuau committed:
Fix: @Optional() FormGroupDirective in OpFormFieldComponent

aleixsuau committed:
Fix: @Optional() FormGroupDirective in OpFormFieldComponent

aleixsuau committed:
Code climate fixes

aleixsuau committed:
noWrapLabel default to false

Commit:
Started adding user custom fields to the ium

aleixsuau committed:
DynamicFormComponent tests

aleixsuau committed:
DynamicFormComponent tests

aleixsuau committed:
Dynamic input test helpers + boolean and text tests

aleixsuau committed:
Dynamic input test helpers + boolean and text tests

Commit:
Import the dynamic-forms module into the common module

aleixsuau committed:
Refactor edit fields to avoid circular dependencies in the dynamic forms

aleixsuau committed:
Refactor edit fields to avoid circular dependencies in the dynamic forms

aleixsuau committed:
Refactor edit fields to avoid circular dependencies in the dynamic forms

aleixsuau committed:
Using DynamicFormsModule in OpenprojectInviteUserModalModule

Commit:
Add formly form

Commit:
Update principal name filter

aleixsuau committed:
Naming fix

aleixsuau committed:
Naming fix

aleixsuau committed:
IntegerInputComponent tests

aleixsuau committed:
IntegerInputComponent tests

aleixsuau committed:
SelectInputComponent tests

aleixsuau committed:
SelectInputComponent tests

aleixsuau committed:
Fix: duplicated identifier on inputs

aleixsuau committed:
Fix: duplicated identifier on inputs

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/36382-multi-selection-for-project-custom-fields-base

oliverguenther committed:
Extract toolbar to be reused for now

Still TBD whether we want to move them right now to the frontend?

oliverguenther committed:
Extract toolbar to be reused for now

Still TBD whether we want to move them right now to the frontend?

oliverguenther committed:
Create new project route and redirect to rails view after saving

oliverguenther committed:
Create new project route and redirect to rails view after saving

Commit:
Dynamic form field is rendering

aleixsuau committed:
fieldsSettingsPipe + hide 'identifier' on projects

aleixsuau committed:
fieldsSettingsPipe + hide 'identifier' on projects

aleixsuau committed:
Handling multi-values (also as links) and passwords

aleixsuau committed:
Handling multi-values (also as links) and passwords

aleixsuau committed:
Handling multi-values (also as links) and passwords

aleixsuau committed:
Some TODOs removed

aleixsuau committed:
Some TODOs removed

aleixsuau committed:
FormattableTextareaInputComponent tests

aleixsuau committed:
FormattableTextareaInputComponent tests

aleixsuau committed:
_isResourceSchema based on parent?.location

aleixsuau committed:
Scope DynamicFieldsService to DynamicFormComponent

aleixsuau committed:
Merged with origin

aleixsuau committed:
Added backend validation method to FormsService

aleixsuau committed:
Added backend validation method to FormsService

aleixsuau committed:
Remove form from DynamicForm when not isStandaloneForm

aleixsuau committed:
Allow multiple form keys to validate

aleixsuau committed:
Remove form from non standalone forms

aleixsuau committed:
Remove duplicated button

Commit:
Merge branch 'feature/36382-multi-selection-for-project-custom-fields-base' into feature/user-custom-field-ium

Commit:
Merge branch 'dev' into feature/user-custom-field-ium

Commit:
Moved to FormGroup input for dynamic form

Commit:
Custom field happy path is done

Commit:
Add explanatory comment to payload structure transformation

Commit:
Merge branch 'dev' into feature/user-custom-field-ium

Commit:
add op-form class to ium steps

Commit:
Add shrinkwrap back in

oliverguenther commented:
![Screenshot from 2021-05-04 13-56-39](https://user-images.githubusercontent.com/459462/117000042-c43e4880-ace0-11eb-911f-2d73ffa4427d.png)

The required CF is shown, but still getting a dynamic-form error on clicking Next after invite

Commit:
Fix test, fix dynamic form resource path

Commit:
gimme a shirnkwrap

Commit:
Remove failing tests

Commit:
Remove another failing test

Commit:
Remove more failing specs

oliverguenther committed:
Reset shrinkwrap

oliverguenther commented:
Moving to the principal component does multiple requests simultaneously (Not sure if caused in this PR)

![Peek 2021-05-04 16-00](https://user-images.githubusercontent.com/459462/117015334-e8eeec00-acf1-11eb-8b8d-fe2be960765c.gif)

oliverguenther commented on frontend/src/app/modules/invite-user-modal/principal/principal.component.ts:
```diff
@@ -93,13 +147,40 @@ export class PrincipalComponent implements OnInit {
+    // The code below transforms the model value as it comes from the dynamic form to the value accepted by the API.
+    // This is not just necessary for submit, but also so that we can reseed the initial values to the payload
+    // when going back to this step after having completed it once.
+    const links = this.customFields!._links || {};
```
Wouldn't the dynamic form need this same helper when transforming the model before submission? Can't we reuse that here? It's really specific to have this in this component

b12f replied:
Yes I agree, and I'm not sure where it would fit best. Perhaps the dynamic form service would be an even better place. To be honest, this feels like something that should just be a pure function and not be tied to any other service or component in particular.

oliverguenther commented:
Functionality-wise, LGTM :+1: I tested most of the user fields with a required flag. Still needs an extension to the feature spec however (I can help you with that if you need)

b12f commented:
> Functionality-wise, LGTM +1 I tested most of the user fields with a required flag. Still needs an extension to the feature spec however (I can help you with that if you need)

Yeah I'd need help with adding specific custom fields. After that I'd be good I think.

b12f commented:
> Moving to the principal component does multiple requests simultaneously (Not sure if caused in this PR)

This is probably old, but I'll take a look

Commit:
Fix double loading of principals

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/user-custom-field-ium

oliverguenther committed:
Add custom field spec

oliverguenther commented:
I added a spec that tests all custom fields. Right now a required float CF is not shown (we don't have a float field yet? @aleixsuau )

I also had to change the code you used to wrap the _links, as a single value list still needs to be rendered in _links and with the href: set. It's apparent that we need to move this to the dynamic form, so that it's not the responsibility of the caller

oliverguenther committed:
Fix spec

oliverguenther committed:
Merge remote-tracking branch 'origin/dev' into feature/user-custom-field-ium

oliverguenther committed:
Forbid Factory.build(:user, member_in_project)

If you use the trait member_in_project(s), the user is implicitly saved
to create the member.

This is very confusing if trying to use required custom fields, as this
will fail with the Member#user_id foreign key being nil, as the user
cannot be saved.

Instead, raise an error when trying to use this factory trait

oliverguenther committed:
Change additional spec factory
